jujugui/static/gui/src/app/state/test-state.js

```diff
@@ -0,0 +1,566 @@
+      }];
+
+      urls.forEach(test => {
+        assert.deepEqual(
+          state.buildState(test.path),
+          test.state,
+          `${test.path} did not properly generate the state object: ` +
+          JSON.stringify(test.state));
+      });
+    });
+
+    it('builds proper state for urls with all sections', () => {
+      const state = new window.jujugui.State({
+        baseURL: 'http://abc.com:123'
+      });
+
+      const urls = [{
+        path: 'http://abc.com:123/u/hatch/staging/haproxy/i/inspector/haproxy/config', // eslint-disable-line max-len
+        state: {
+          user: 'hatch/staging',
+          store: 'haproxy',
+          gui: { inspector: 'haproxy/config' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/frankban/production/ghost/xenial/i/inspector/haproxy/config', // eslint-disable-line max-len
+        state: {
+          user: 'frankban/production',
+          store: 'ghost/xenial',
+          gui: { inspector: 'haproxy/config' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/hatch/staging/ghost/42/i/inspector/haproxy/config', // eslint-disable-line max-len
+        state: {
+          user: 'hatch/staging',
+          store: 'ghost/42',
+          gui: { inspector: 'haproxy/config' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/frankban/production/django/bundle/47/i/machines', // eslint-disable-line max-len
+        state: {
+          user: 'frankban/production',
+          store: 'django/bundle/47',
+          gui: { machines: '' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/hatch/staging/u/frankban/django/i/applications/inspector/ghost/', // eslint-disable-line max-len
+        state: {
+          user: 'hatch/staging',
+          store: 'frankban/django',
+          gui: { applications: '', inspector: 'ghost' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/frankban/production/u/hatch/mongodb/xenial/i/machines', // eslint-disable-line max-len
+        state: {
+          user: 'frankban/production',
+          store: 'hatch/mongodb/xenial',
+          gui: { machines: '' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/hatch/staging/u/hatch/mongodb/47/i/applications/inspector/ghost/', // eslint-disable-line max-len
+        state: {
+          user: 'hatch/staging',
+          store: 'hatch/mongodb/47',
+          gui: { applications: '', inspector: 'ghost' }
+        }
+      }, {
+        path: 'http://abc.com:123/u/frankban/production/u/frankban/django/bundle/0/i/applications', // eslint-disable-line max-len
+        state: {
+          user: 'frankban/production',
+          store: 'frankban/django/bundle/0',
+          gui: { applications: '' }
+        }
+      }];
+
+      urls.forEach(test => {
+        assert.deepEqual(
+          state.buildState(test.path),
+          test.state,
+          `${test.path} did not properly generate the state object: ` +
+          JSON.stringify(test.state));
```

Comment:
These tests are great Jeff, thanks!
I don't see any of the error cases here. maybe this lib should handle apparent 404 as well?
For instance, could you add tests for all the following page?
```
'/u',
    '/about/wat',
    '/no-such/i',
    '/django/bundle/42/wat',
    '/u/frankban/u',
    '/u/frankban/u/haproxy',
    '/u/hatch/staging/u/',
    '/u/hatch/staging/u/hatch',
    '/u/frankban/production/u/frankban/i/applications',
    '/u/frankban/u/frankban/haproxy',
    '/ghost/u/frankban/ghost',
    '/u/hatch/staging/u/hatch/ghost/trusty/42/wat',
    '/u/hatch/staging/u/hatch/',
    '/u/hatch/staging/hatch/u/wat',
```
They are all invalid, and we can detect that with a static check in the state library.